### PR TITLE
[WIP] Refactoring reservations

### PR DESCRIPTION
**Pull request by Nekotekina**
This is only PoC, may not work properly or have degraded performance. I want to figure this out since it opens a path towards various potential improvements.

### CODE REVIEW

**Comment by RainbowCookie32** [comment]
Did some testing, found some issues.

GoW 3 likes to freeze on Sony's or Santa Monica's logos after booting the game, or playing them at a few seconds per frame. After I did manage to get to the menu, performance seems to be lower on this PR compared to Master.

Master:
gow3_master

This PR:
gow3_pr

---

Both Killzone 2's Demo and GTA V got stuck after boot on this PR. Tried a few times, but I haven't managed to get ingame on either.

Killzone 2 gets stuck on a black screen:
kz2d_pr
[Log from PR](https://github.com/RPCS3/rpcs3/files/6216472/RPCS3.log.gz)
[Log from Master](https://github.com/RPCS3/rpcs3/files/6216483/RPCS3.log.gz)

and GTA V plays its logos video for a bit and freezes:
gta5_pr
[Log from PR](https://github.com/RPCS3/rpcs3/files/6216478/RPCS3.log.gz)
[Log from Master](https://github.com/RPCS3/rpcs3/files/6216495/RPCS3.log.gz)

---

Persona 5's performance seems to be the same as Master.

Master:
p5_master

This PR:
p5_pr

File: rpcs3/Emu/Cell/SPUThread.h [hunk 1]
```diff
@@ -765,12 +771,11 @@ class spu_thread : public cpu_thread
 	u64 stx = 0; // Succeeded transactions (pure counters)
 
 	u64 last_ftsc = 0;
-	u64 last_ftime = 0;
-	u32 last_faddr = 0;
 	u64 last_fail = 0;
 	u64 last_succ = 0;
+	u32 last_faddr = 0;
 
-	u64 mfc_dump_idx = 0;
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPUThread.h:773]: u64 was to avoid overflows over time.

File: rpcs3/Emu/Cell/PPUThread.h [hunk 2]
```diff
@@ -241,7 +241,6 @@ class ppu_thread : public cpu_thread
 	u32 jm_mask = 0x7f80'0000;
 
 	u32 raddr{0}; // Reservation addr
-	u64 rtime{0};
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUThread.h:244]: I suggest keeping it for accurate reservations setting.

File: rpcs3/Emu/Cell/SPUThread.h (hunk 1 above)
Comment (Nekotekina) [reply rpcs3/Emu/Cell/SPUThread.h:773]: Okay, I'll fix it.

**Comment by Nekotekina** [comment]
@RainbowCookie32 did you test with TSX or without?

**Comment by dio-gh** [comment]
He runs AMD.

File: rpcs3/Emu/Cell/PPUThread.cpp [hunk 3]
```diff
@@ -2003,23 +1978,21 @@ static bool ppu_store_reservation(ppu_thread& ppu, u32 addr, u64 reg_value)
 		if (new_data == old_data)
 		{
 			ppu.last_faddr = 0;
-			return res.compare_and_swap_test(rtime, rtime + 128);
```
Comment (elad335) [inline rpcs3/Emu/Cell/PPUThread.cpp:2006]: It won't work for pure reservation notification where one waits on SPU LR and PPU starts dummy updating reservation with no value changes, keep this code under PPU accirate reservations.

**Comment by RainbowCookie32** [comment]
Indeed. Ryzen 7 1700, so no TSX.

**Comment by Nekotekina** [comment]
Heavily outdated.

**Pull request closed without merging**